module/documents/actor/actor.mjs

```diff
@@ -1359,191 +1359,208 @@ export default class Actor5e extends SystemDocumentMixin(Actor) {
+  /**
+   * Shared rolling functionality between skill & tool checks.
+   * @param {"skill"|"tool"} type                                Type of roll.
+   * @param {Partial<SkillToolRollProcessConfiguration>} config  Configuration information for the roll.
+   * @param {Partial<SkillToolRollDialogConfiguration>} dialog   Configuration for the roll dialog.
+   * @param {Partial<BasicRollMessageConfiguration>} message     Configuration for the roll message.
+   * @returns {Promise<D20Roll[]|null>}                          A Promise which resolves to the created Roll instance.
+   */
+  async #rollSkillTool(type, config={}, dialog={}, message={}) {
+    let oldFormat = false;
+    const name = type === "skill" ? "Skill" : "ToolCheck";
+
+    // Handle deprecated calling pattern
+    if ( foundry.utils.getType(config) !== "Object" ) {
+      foundry.utils.logCompatibilityWarning(
+        `The \`roll${name}\` method on Actor5e now takes roll, dialog, and message config objects as parameters.`,
+        { since: "DnD5e 4.1", until: "DnD5e 4.5" }
+      );
+      oldFormat = true;
+      const oldConfig = dialog;
+      config = { [type]: config };
+      if ( oldConfig.ability ) config.ability = oldConfig.ability;
+      if ( oldConfig.bonus ) config.bonus = oldConfig.bonus;
+      dialog = {};
+      _applyDeprecatedD20Configs(config, dialog, message, oldConfig);
     }
 
-    // Global ability check bonus.
-    if ( globalBonuses.check ) {
-      parts.push("@checkBonus");
-      data.checkBonus = Roll.replaceFormulaData(globalBonuses.check, data);
+    const skill = this.system.skills?.[config.skill];
+    const skillConfig = CONFIG.DND5E.skills[config.skill];
+    const tool = this.system.tools?.[config.tool];
+    const toolConfig = CONFIG.DND5E.tools[config.tool];
+    if ( ((type === "skill") && !skillConfig) || ((type === "tool") && !toolConfig) ) {
+      return this.rollAbilityTest(config, dialog, message);
     }
 
-    // Ability-specific check bonus.
-    if ( ability?.bonuses.check ) data.abilityCheckBonus = Roll.replaceFormulaData(ability.bonuses.check, data);
-    else data.abilityCheckBonus = 0;
+    const relevant = type === "skill" ? skill : tool;
+    const rollData = this.getRollData();
+    const initialRoll = config.rolls?.pop();
+    const defaultAbility = relevant?.ability ?? (type === "skill" ? skillConfig.ability : toolConfig.ability);
 
-    // Tool-specific check bonus.
-    if ( tool?.bonuses.check || options.bonus ) {
-      parts.push("@toolBonus");
-      const bonus = [];
-      if ( tool?.bonuses.check ) bonus.push(Roll.replaceFormulaData(tool.bonuses.check, data));
-      if ( options.bonus ) bonus.push(Roll.replaceFormulaData(options.bonus, data));
-      data.toolBonus = bonus.join(" + ");
-    }
+    const hasRemarkableAthlete = this.getFlag("dnd5e", "remarkableAthlete")
+      && (game.settings.get("dnd5e", "rulesVersion") === "legacy");
+    const buildConfig = (process, config, formData, index) => {
```

Comment:
Rather than a callback, could we have our `SkillToolRollConfiguration` subclass override `_buildRoll` and call a well-known method on the Actor? Something like `_prepareSkillCheck`/`_prepareToolCheck` maybe?

Reply:
That does add some complications to parts because right now we rely on bound variables for some of the calculations (relevant skill, default ability, and roll data so we only have to prepare it once, plus initial roll data because that is passed in to the initial configuration function). I could perhaps split this out into its own method but bind these values into it before passing it to the roll config:

```javascript
const buildConfig = this._buildSkillToolConfig.bind(this, { defaultAbility, initialRoll, relevant, rollData });
```

Reply:
Binding is an option. It's not an issue to recompute things like `rollData`, or `hasRemarkableAthlete` inside the method, since it's only called a single time as part of user interaction.

This is how far I got with just a pure method-based solution:

```js
// SkillToolRollConfigurationDialog
// Add 'type' and 'actor' to dialog options
/**
 * @typedef {BasicRollConfigurationDialogOptions} SkillToolRollConfigurationDialogOptions
 * @property {boolean} chooseAbility  Should the ability be selectable?
 * @property {"skill"|"tool"} type    The type of roll.
 * @property {Actor5e} actor          The Actor making the roll.
 */
_buildConfig(config, formData, index) {
  super._buildConfig(config, formData, index);
  this.options.actor._prepareSkillTool(this.options.type, this.config, config, formData);
}
```

```js
// Actor5e
async #rollSkillTool(type, config={}, dialog={}, message={}) {
  // [snip]
  const rollConfig = { /* ... */ };
  this._prepareSkillTool(type, rollConfig, rollConfig.rolls[0], null);
  const dialogConfig = foundry.utils.mergeObject({
    applicationClass: SkillToolRollConfigurationDialog,
    options: { type, chooseAbility: true, actor: this }
  }, dialog);
}

/**
 * @protected
 */
_prepareSkillTool(type, process, config, formData) {
  const rollData = this.getRollData();
  // Should not need to derive defaultAbility here since process.ability is already set to that value.
  const abilityId = formData?.get("ability") ?? process.ability;
  const ability = this.system.abilities?.[abilityId];
  const skill = this.system.skills?.[process.skill];
  const tool = this.system.tools?.[process.tool];
  const relevant = type === "skill" ? skill : tool;
  const hasRemarkableAthlete = this.getFlag("dnd5e", "remarkableAthlete")
    && (game.settings.get("dnd5e", "rulesVersion") === "legacy");

  let prof = process.prof ?? relevant?.prof;
  if ( prof && hasRemarkableAthlete ) {
    const validAbilities = CONFIG.DND5E.characterFlags.remarkableAthlete.abilities;
    if ( validAbilities.includes(abilityId) && (prof.multiplier <= .5) && (prof.rounding === "down") ) {
      prof = prof.clone({ multiplier: .5, roundDown: false });
    } else if ( !validAbilities.includes(abilityId) && (prof.multiplier === .5) && (prof.rounding === "up") ) {
      if ( this.getFlag("dnd5e", "jackOfAllTrades") ) prof = prof.clone({ roundDown: true });
      else prof = prof.clone({ multiplier: 0 });
    }
  }

  const { parts, data } = CONFIG.Dice.BasicRoll.constructParts({
    mod: ability?.mod,
    prof: prof?.hasProficiency ? prof.term : null,
    [`${config[type]}Bonus`]: relevant?.bonuses?.check,
    extraBonus: process.bonus,
    [`${abilityId}CheckBonus`]: ability?.bonuses?.check,
    [`${type}Bonus`]: this.system.bonuses?.abilities?.[type],
    abilityCheckBonus: this.system.bonuses?.abilities?.check,
    situational: config.data?.situational
  }, { ...rollData });
  const options = config.options ?? {};

  this.addRollExhaustion(parts, data);
  config.parts = parts;
  config.data = data;
  config.data.abilityId = abilityId;
  config.options = options;
}
```

I think passing `type` to the `SkillToolRollConfigurationDialog` is pretty reasonable. The `SkillToolRollConfigurationDialog` should know whether it's a skill roll or a tool roll. But I don't love having to pass the `Actor` really. So I think this exercise has me leaning more towards the binding solution, where we bind the `Actor` as `this` and then the roll dialog doesn't need to know.

I would like to understand a bit better what we are doing with `initialRoll` though since that's the only part missing from the above. Is it a holdover from the old API? I don't think we use this at all internally, so is this something related to module integration?

Reply:
`initialRoll` is there to better support calling `rollSkill` or `rollToolCheck` with default options. So if you call:

```javascript
actor.rollSkill({ roll: [{ parts: ["1d4"] }] });
```

That will add the provided `1d4` part to the roll that is automatically constructed, so you end up with `1d20 + @mod + 1d4`. Same with the data and options provided there, allowing you to override what would be automatically calculated.

For all other parts we just append a new roll, so if we didn't handle initial roll we'd end up with one roll `1d20 + @mod`, and a second roll `1d20 + @mod + 1d4` (that's still possible in this system, but you have to pass multiple options to `roll` in the first object).

Reply:
OK I think that makes sense. What confused me was that we call `config.rolls?.pop()`, but it sounds like from what you described that we should be doing `config.rolls?.shift()`, no? Also, why do we need to do this merge every time in the callback? If these options are supposed to be merged with the generated roll, can't we do it just once when we assemble `rollConfig`? I can see that we are supplying some roll options already to the first roll when we assemble `rollConfig`.

Reply:
Aw yeah, shift is correct. We do this every preparation because we need to merge these in after the default roll parts, data, and options are prepared since these should take priority over those options, and those underlying bits might change every time the user makes a choice.